src/librustc/infer/outlives/verify.rs

```diff
@@ -74,10 +74,18 @@ impl<'cx, 'gcx, 'tcx> VerifyBoundCx<'cx, 'gcx, 'tcx> {
 
     /// Given a projection like `T::Item`, searches the environment
     /// for where-clauses like `T::Item: 'a`. Returns the set of
-    /// regions `'a` that it finds.  This is a "conservative" check --
-    /// it may not find all applicable bounds, but all the bounds it
-    /// returns can be relied upon.
-    pub fn projection_declared_bounds_from_env(
+    /// regions `'a` that it finds.
+    ///
+    /// This is an "approximate" check -- it may not find all
+    /// applicable bounds, and not all the bounds it returns can be
+    /// relied upon. In particular, this check ignores region
+    /// identity.  So, for example, if we have `<T as
+    /// Trait<'0>>::Item` where `'0` is a region variable, and the
+    /// user has `<T as Trait<'a>>::Item: 'b` in the environment, then
+    /// the clause from the environment only applies if `'0 = 'a`,
+    /// which we don't know yet. But we would still include `'b` in
+    /// this list.
```

Comment:
I'm going through the PR commit-by-commit, so this question may get answered later on, but I don't want to forget to ask it: What is the intended use of this check, where it sounds like it is incomplete on two axes...?

My current hypothesis is that this check is used to provide diagnostics in scenarios where we are comfortable giving feedback suggesting the user *might* need to satisfy some constraint, when the reality is that they might be able to sidestep the constraint... That probably didn't make sense. let me try again

Here's a different version of my question: You have given a concrete example where the list returned by this function would include `'b`, even though that bound may not apply, since it depends on the whether it is true that `'0 = 'a`. So: Do you have an example of a diagnostic where as a user one can *see* the effect of `'b` being on this list?

Reply:
My question might be answered by cebf83502fd7c57dd2a7d1481b00d93ec03bab0e

Reply:
(Though commit cebf83502fd7c57dd2a7d1481b00d93ec03bab0e only has tests being updated that are observing internal region representations... I *think* some of tests still show side-effects that end-users can observe without using `-Z` flags, but I *hope* that by the time I get to the end of the commit series, there will be a concrete test that demonstrates the effect without `-Z` flags.)